### PR TITLE
[16647] Use shared_ptr for internal singletons

### PR DESCRIPTION
**Pull request by MiguelCompany**
## Description

The changes include using shared pointers for the internal singletons of RTPSDomainImpl and LogResources, and keeping a reference to both on DomainParticipantFactory.

I consider the changes ABI safe, because the only ABI incompatibility is on the size of DomainParticipantFactory, where two shared pointers are added at the end. Since the factory cannot be directly constructed by the user (and thus cannot be instantiated on the stack), there's no way of having ABI issues. I already checked that by building the examples before the changes, and updating the shared library after the changes.

This is related to #2537, and can be considered a follow-up of #2894.

@Mergifyio backport 2.8.x

## Contributor Checklist
- [x] Commit messages follow the project guidelines. 
- [x] The code follows the style guidelines of this project. 
- **N/A** Tests that thoroughly check the new feature have been added/Regression tests checking the bug and its fix have been added. 
- [x] Any new/modified methods have been properly documented using Doxygen. 
- [ ] Fast DDS test suite has been run locally. 
- [x] Changes are ABI compatible. 
- [x] Changes are API compatible. 
- [ ] Documentation builds and tests pass locally. 
- [ ] New feature has been added to the `versions.md` file (if applicable).
- **N/A** New feature has been documented/Current behavior is correctly described in the documentation. 

## Reviewer Checklist
- [ ] Check contributor checklist is correct.
- [ ] Check CI results: changes do not issue any warning.
- [ ] Check CI results: failing tests are unrelated with the changes.

### CODE REVIEW

**Comment by MiguelCompany** [comment]
@richiprosima Please test this.

**Comment by MiguelCompany** [comment]
@richiprosima Please test this

**Comment by MiguelCompany** [comment]
@richiprosima Please test this

**Comment by MiguelCompany** [comment]
@richiprosima Please test this

**Comment by MiguelCompany** [comment]
@richiprosima Please test this

**Comment by MiguelCompany** [comment]
@richiprosima Please test this

**Comment by MiguelCompany** [comment]
@richiprosima Please test this

**Comment by MiguelCompany** [comment]
@richiprosima Please test this

**Comment by MiguelCompany** [comment]
@richiprosima Please test this

**Review by jsantiago-eProsima — CHANGES_REQUESTED** [review]

File: src/cpp/utils/shared_memory/BoostAtExitRegistry.hpp [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+        std::lock_guard<std::mutex> _(mtx_);
+        registered_functions_.push_back(f);
+    }
+
+private:
+
+    ~BoostAtExitRegistry()
+    {
+        // Get a local copy of the collection
+        std::vector<void (*)()> collection;
```
[the remark this answers is not in the document]
Comment (jsantiago-eProsima) [reply src/cpp/utils/shared_memory/BoostAtExitRegistry.hpp:65]: After reviewing the behavior when registering a function via std::atexit from insida a function registered previously, it seems that it is allowed, with the functions being stored in a stack.

File: src/cpp/rtps/RTPSDomainImpl.hpp [hunk 2]
```diff
@@ -145,7 +208,26 @@ class RTPSDomainImpl
...
+
+    std::mutex m_mutex;
+
+    std::atomic<uint32_t> m_maxRTPSParticipantID;
+
+    std::vector<t_p_RTPSParticipant> m_RTPSParticipants;
+
+    std::set<uint32_t> m_RTPSParticipantIDs;
+
+    FileWatchHandle file_watch_handle_;
```
[the remark this answers is not in the document]
Comment (MiguelCompany) [reply src/cpp/rtps/RTPSDomainImpl.hpp:230]: The whole point of this PR is to remove static objects to avoid destruction order fiascos

**Review by jsantiago-eProsima — APPROVED** [review]
LGTM

**Comment by jsantiago-eProsima** [comment]
@richiprosima please test this

**Review by jsantiago-eProsima — APPROVED** [review]
This still LGTM.

**Comment by MiguelCompany** [comment]
@Mergifyio backport 2.8.x

**Pull request merged**